File: android/app/src/main/kotlin/com/stockbook/app/photos/PhotoViews.kt
package com.stockbook.app.photos

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.gestures.detectTransformGestures
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.produceState
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.stockbook.app.design.GhostButton
import com.stockbook.app.design.Glyph
import com.stockbook.app.design.Icon
import com.stockbook.app.design.IconButton
import com.stockbook.app.design.Metrics
import com.stockbook.app.design.Nocturne
import com.stockbook.app.design.NocturneType
import com.stockbook.app.design.hairline
import com.stockbook.core.text.Strings
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * The strip of photographs under a bill.
 *
 * Thumbnails rather than a list of file names, because the whole point of a
 * photograph is that it is recognised at a glance. Tapping one opens it full
 * screen; there is no second level of navigation for a picture of a receipt.
 */
@Composable
fun PhotoStrip(
    ids: List<String>,
    strings: Strings,
    modifier: Modifier = Modifier,
    onOpen: (String) -> Unit
) {
    if (ids.isEmpty()) return

    LazyRow(
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        modifier = modifier.fillMaxWidth()
    ) {
        items(ids, key = { it }) { id ->
            PhotoThumbnail(id = id, strings = strings, onClick = { onOpen(id) })
        }
    }
}

@Composable
private fun PhotoThumbnail(id: String, strings: Strings, onClick: () -> Unit) {
    val context = LocalContext.current
    val photos = remember(context) { PhotoStore(context) }

    // Decoding is file work, and file work on the frame thread is how a strip of
    // four pictures drops frames while it draws.
    val bitmap by produceState<android.graphics.Bitmap?>(null, id) {
        value = withContext(Dispatchers.IO) { photos.bitmap(id, edge = 320) }
    }

    Box(
        contentAlignment = Alignment.Center,
        modifier = Modifier
            .size(width = 76.dp, height = 92.dp)
            .clip(RoundedCornerShape(Metrics.controlRadius))
            .background(Nocturne.surface)
            .hairline(radius = Metrics.controlRadius)
            .clickable(onClick = onClick)
    ) {
        val image = bitmap
        if (image != null) {
            Image(
                bitmap = image.asImageBitmap(),
                contentDescription = strings.billPhotos,
                contentScale = ContentScale.Crop,
                modifier = Modifier.fillMaxSize()
            )
        } else {
            // A book can arrive on a phone ahead of its pictures. Saying so is
            // the honest answer; an empty square is one the owner has to guess at.
            MissingPhoto(strings)
        }
    }
}

@Composable
private fun MissingPhoto(strings: Strings) {
    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center,
        modifier = Modifier.padding(horizontal = 6.dp)
    ) {
        Glyph(Icon.items, size = 14.dp, tint = Nocturne.neutral500)
        Spacer(Modifier.height(4.dp))
        Text(
            strings.photoNotOnThisPhone,
            style = NocturneType.meta,
            color = Nocturne.neutral500,
            textAlign = TextAlign.Center
        )
    }
}

/**
 * One photograph, filling the screen.
 *
 * Pinch to zoom and drag to move, because the reason to open a picture of a bill
 * is almost always to read something small on it. Nothing else: no captions, no
 * editing, no filters. It is a photograph of a piece of paper.
 */
@Composable
fun PhotoViewer(
    id: String,
    strings: Strings,
    onShare: () -> Unit,
    onRemove: () -> Unit,
    onClose: () -> Unit
) {
    val context = LocalContext.current
    val photos = remember(context) { PhotoStore(context) }
    val bitmap by produceState<android.graphics.Bitmap?>(null, id) {
        value = withContext(Dispatchers.IO) { photos.bitmap(id) }
    }

    var scale by remember { mutableFloatStateOf(1f) }
    var offsetX by remember { mutableFloatStateOf(0f) }
    var offsetY by remember { mutableFloatStateOf(0f) }
    var confirmingRemoval by remember { mutableStateOf(false) }

    Column(modifier = Modifier.fillMaxWidth()) {
        Row(verticalAlignment = Alignment.CenterVertically, modifier = Modifier.fillMaxWidth()) {
            Text(
                strings.billPhotos,
                style = NocturneType.inter(15.0),
                color = Nocturne.text,
                modifier = Modifier.weight(1f)
            )
            IconButton(Icon.share, onClick = onShare, size = 16.dp, contentDescription = strings.share)
            Spacer(Modifier.width(4.dp))
            IconButton(Icon.close, onClick = onClose, size = 16.dp, contentDescription = strings.done)
        }

        Spacer(Modifier.height(10.dp))

        Box(
            contentAlignment = Alignment.Center,
            modifier = Modifier
                .fillMaxWidth()
                .height(420.dp)
                .clip(RoundedCornerShape(Metrics.controlRadius))
                .background(Color.Black)
        ) {
            val image = bitmap
            if (image != null) {
                Image(
                    bitmap = image.asImageBitmap(),
                    contentDescription = strings.billPhotos,
                    contentScale = ContentScale.Fit,
                    modifier = Modifier
                        .fillMaxSize()
                        .graphicsLayer(
                            scaleX = scale,
                            scaleY = scale,
                            translationX = offsetX,
                            translationY = offsetY
                        )
                        .pointerInput(id) {
                            detectTransformGestures { _, pan, zoom, _ ->
                                // Floored at 1: a photograph smaller than its own
                                // frame is a picture nobody asked for. Capped at
                                // 5, which is past the point where a stored
                                // photograph has any more detail to give.
                                scale = (scale * zoom).coerceIn(1f, 5f)
                                if (scale > 1f) {
                                    offsetX += pan.x
                                    offsetY += pan.y
                                } else {
                                    // Back to the middle on the way out, so the
                                    // next pinch does not start somewhere the
                                    // owner did not leave it.
                                    offsetX = 0f
                                    offsetY = 0f
                                }
                            }
                        }
                )
            } else {
                MissingPhoto(strings)
            }
        }

        Spacer(Modifier.height(10.dp))

        GhostButton(
            if (confirmingRemoval) strings.tapAgainToRemove else strings.removePhoto,
            onClick = {
                if (confirmingRemoval) onRemove() else confirmingRemoval = true
            },
            tint = if (confirmingRemoval) Nocturne.accent400 else Nocturne.neutral500,
            modifier = Modifier.fillMaxWidth()
        )
    }
}

File: ios/Stockbook/Features/Bills/PhotoViews.swift
import SwiftUI
import PhotosUI

/// The strip of photographs under a bill.
///
/// Thumbnails rather than a list of file names, because the whole point of a
/// photograph is that it is recognised at a glance. Tapping one opens it full
/// screen; there is no second level of navigation for a picture of a receipt.
struct PhotoStrip: View {
    let ids: [String]
    let onOpen: (String) -> Void

    var body: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 8) {
                ForEach(ids, id: \.self) { id in
                    PhotoThumbnail(id: id)
                        .onTapGesture { onOpen(id) }
                }
            }
        }
    }
}

private struct PhotoThumbnail: View {
    let id: String

    @State private var image: UIImage?

    var body: some View {
        ZStack {
            RoundedRectangle(cornerRadius: Metrics.controlRadius, style: .continuous)
                .fill(Nocturne.surface)

            if let image {
                Image(uiImage: image)
                    .resizable()
                    .aspectRatio(contentMode: .fill)
            } else {
                // A book can arrive on a phone ahead of its pictures. Saying so
                // is the honest answer; an empty square is one the owner has to
                // guess at.
                MissingPhoto()
            }
        }
        .frame(width: 76, height: 92)
        .clipShape(RoundedRectangle(cornerRadius: Metrics.controlRadius, style: .continuous))
        .hairline(radius: Metrics.controlRadius)
        .task(id: id) { image = await PhotoLoader.load(id: id, edge: 320) }
    }
}

private struct MissingPhoto: View {
    var body: some View {
        VStack(spacing: 4) {
            Glyph(Icon.items, size: 14).foregroundStyle(Nocturne.neutral500)
            Text(Loc.photoNotOnThisPhone)
                .nocturneText(.meta)
                .multilineTextAlignment(.center)
        }
        .padding(.horizontal, 6)
    }
}

/// One photograph, filling the sheet.
///
/// Pinch to zoom and drag to move, because the reason to open a picture of a bill
/// is almost always to read something small on it. Nothing else: no captions, no
/// editing, no filters. It is a photograph of a piece of paper.
struct PhotoViewer: View {
    let id: String
    let onRemove: () -> Void
    let onClose: () -> Void

    @State private var image: UIImage?
    @State private var scale: CGFloat = 1
    @State private var offset: CGSize = .zero
    @State private var confirmingRemoval = false

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            HStack {
                Text(Loc.billPhotos)
                    .font(NocturneType.inter(15))
                    .foregroundStyle(Nocturne.text)
                Spacer()
                // The file itself, not a rendering of it. Whatever the owner
                // picks in the share sheet receives a JPEG of the paper bill,
                // and the app makes no network call to send it.
                ShareLink(item: PhotoStore().url(id: id)) { Glyph(Icon.share, size: 16) }
                    .accessibilityLabel(Loc.share)
                Button(action: onClose) { Glyph(Icon.close, size: 16) }
                    .buttonStyle(.plain)
                    .padding(.leading, 12)
                    .accessibilityLabel(Loc.done)
            }
            .foregroundStyle(Nocturne.text)
            .padding(.bottom, 10)

            ZStack {
                Color.black
                if let image {
                    Image(uiImage: image)
                        .resizable()
                        .aspectRatio(contentMode: .fit)
                        .scaleEffect(scale)
                        .offset(offset)
                        .gesture(zoom.simultaneously(with: drag))
                } else {
                    MissingPhoto()
                }
            }
            .frame(height: 420)
            .clipShape(RoundedRectangle(cornerRadius: Metrics.controlRadius, style: .continuous))

            Button(confirmingRemoval ? Loc.tapAgainToRemove : Loc.removePhoto) {
                if confirmingRemoval { onRemove() } else { confirmingRemoval = true }
            }
            .buttonStyle(GhostButtonStyle(
                fontSize: 12.5,
                tint: confirmingRemoval ? Nocturne.accent400 : Nocturne.neutral500
            ))
            .frame(maxWidth: .infinity)
            .padding(.top, 10)
        }
        .task(id: id) { image = await PhotoLoader.load(id: id) }
    }

    /// Floored at 1: a photograph smaller than its own frame is a picture nobody
    /// asked for. Capped at 5, past which a stored photograph has no more detail
    /// to give.
    private var zoom: some Gesture {
        MagnifyGesture()
            .onChanged { scale = max(1, min(5, $0.magnification)) }
            .onEnded { _ in
                if scale <= 1 {
                    // Back to the middle on the way out, so the next pinch does
                    // not start somewhere the owner did not leave it.
                    scale = 1
                    offset = .zero
                }
            }
    }

    private var drag: some Gesture {
        DragGesture()
            .onChanged { value in
                guard scale > 1 else { return }
                offset = value.translation
            }
    }
}

/// Reads a stored photograph off the main thread.
///
/// `edge` is what the caller is about to draw into. A row of thumbnails asking
/// for full-size images would hold tens of megabytes to draw a strip of postage
/// stamps, so the decoder is told what is actually wanted and does the throwing
/// away itself.
enum PhotoLoader {
    static func load(id: String, edge: Int = PhotoPolicy.maxEdge) async -> UIImage? {
        let url = PhotoStore().url(id: id)
        return await Task.detached(priority: .userInitiated) {
            guard let source = CGImageSourceCreateWithURL(url as CFURL, nil) else { return nil }
            let options: [CFString: Any] = [
                kCGImageSourceCreateThumbnailFromImageAlways: true,
                kCGImageSourceCreateThumbnailWithTransform: true,
                kCGImageSourceThumbnailMaxPixelSize: edge
            ]
            guard let cg = CGImageSourceCreateThumbnailAtIndex(source, 0, options as CFDictionary) else {
                return nil
            }
            return UIImage(cgImage: cg)
        }.value
    }
}

/// The camera, wrapped just enough to hand back one photograph.
///
/// `PhotosPicker` covers the gallery natively; the camera has no SwiftUI
/// equivalent, so this is the one place a `UIViewController` shows through. It
/// needs `NSCameraUsageDescription`, which is the single thing this feature asks
/// the phone for on either platform.
struct CameraSheet: UIViewControllerRepresentable {
    let onCaptured: (Data?) -> Void

    func makeUIViewController(context: Context) -> UIImagePickerController {
        let picker = UIImagePickerController()
        picker.sourceType = .camera
        picker.delegate = context.coordinator
        return picker
    }

    func updateUIViewController(_ controller: UIImagePickerController, context: Context) {}

    func makeCoordinator() -> Coordinator { Coordinator(onCaptured: onCaptured) }

    final class Coordinator: NSObject, UIImagePickerControllerDelegate, UINavigationControllerDelegate {
        private let onCaptured: (Data?) -> Void

        init(onCaptured: @escaping (Data?) -> Void) { self.onCaptured = onCaptured }

        func imagePickerController(
            _ picker: UIImagePickerController,
            didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey: Any]
        ) {
            let image = info[.originalImage] as? UIImage
            // Full quality here on purpose: `PhotoStore` is what decides how
            // large a kept photograph is, and re-compressing on the way in would
            // throw detail away twice.
            onCaptured(image?.jpegData(compressionQuality: 1))
        }

        func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
            onCaptured(nil)
        }
    }

    /// Whether this phone has one at all. A simulator does not.
    static var isAvailable: Bool {
        UIImagePickerController.isSourceTypeAvailable(.camera)
    }
}
